ocaml/xapi/xapi_psr.ml

```diff
@@ -145,11 +167,214 @@ functor
         )
 
     let start pool_secrets ~master ~members =
-      match Impl.retrieve_checkpoint () |> Option.map checkpoint_of_string with
-      | None | Some No_checkpoint ->
-          go pool_secrets master members No_checkpoint
-      | Some checkpoint ->
-          go (Impl.retrieve ()) master members checkpoint
+      try
+        match
+          Impl.retrieve_checkpoint () |> Option.map checkpoint_of_string
+        with
+        | None | Some No_checkpoint ->
+            go pool_secrets master members No_checkpoint
+        | Some checkpoint ->
+            go (Impl.retrieve ()) master members checkpoint
+      with e ->
+        (* _in theory_ save_checkpoint or backup could fail, so
+           catch that here. however we don't expect this to happen *)
+        D.error "PSR.start: unexpected error: %s" (Printexc.to_string e) ;
+        raise
+          Api_errors.(
+            Server_error (internal_error, ["PSR.start: unexpected error"]))
+  end
+
+let perm = 0o640
+
+let cleanup_internal ~additional_files_to_remove () =
+  if List.length !Xapi_globs.pool_secrets <= 1 then
+    D.info "xapi_psr.ml:cleanup_internal: already cleaned up"
+  else
+    (* if we don't remove pool secret backups, they will be loaded into memory when xapi starts *)
+    let files_to_remove =
+      List.append additional_files_to_remove
+        [old_pool_secret_backup_path; new_pool_secret_backup_path]
+    in
+    List.iter
+      (fun path -> if Sys.file_exists path then Sys.remove path)
```

Comment:
In general it is better to avoid a test and just do the action and deal with the failure - this avoids race conditions.